src/device_memory_tracker.cc

```diff
@@ -0,0 +1,288 @@
+        std::lock_guard<std::mutex> lk(mtx_);
+        auto it = activity_to_memory_usage_.find(memory_record->correlationId);
+        if (it != activity_to_memory_usage_.end()) {
+          usage = reinterpret_cast<TRITONBACKEND_CuptiTracker*>(it->second);
+          activity_to_memory_usage_.erase(it);
+        }
+      }
+      // Ignore memory record that is not associated with a TRITONBACKEND_CuptiTracker
+      // object
+      if (usage != nullptr) {
+        const bool is_allocation =
+            (memory_record->memoryOperationType ==
+             CUPTI_ACTIVITY_MEMORY_OPERATION_TYPE_ALLOCATION);
+        const bool is_release =
+            (memory_record->memoryOperationType ==
+             CUPTI_ACTIVITY_MEMORY_OPERATION_TYPE_RELEASE);
+        if (is_allocation || is_release) {
+          switch (memory_record->memoryKind) {
+            case CUPTI_ACTIVITY_MEMORY_KIND_DEVICE: {
+              if (memory_record->deviceId >= usage->cuda_array_len_) {
+                usage->good_record_ = false;
+                break;
+              }
+              if (is_allocation) {
+                usage->cuda_byte_size_[memory_record->deviceId] +=
+                    memory_record->bytes;
+              } else {
+                usage->cuda_byte_size_[memory_record->deviceId] -=
+                    memory_record->bytes;
+              }
+              break;
+            }
+            case CUPTI_ACTIVITY_MEMORY_KIND_PINNED: {
+              if (memory_record->deviceId >= usage->pinned_array_len_) {
+                usage->good_record_ = false;
+                break;
+              }
+              if (is_allocation) {
+                usage->pinned_byte_size_[memory_record->deviceId] +=
+                    memory_record->bytes;
+              } else {
+                usage->pinned_byte_size_[memory_record->deviceId] -=
+                    memory_record->bytes;
+              }
+              break;
+            }
+            case CUPTI_ACTIVITY_MEMORY_KIND_PAGEABLE: {
+              if (memory_record->deviceId >= usage->system_array_len_) {
+                usage->good_record_ = false;
+                break;
+              }
+              if (is_allocation) {
+                usage->system_byte_size_[memory_record->deviceId] +=
+                    memory_record->bytes;
+              } else {
+                usage->system_byte_size_[memory_record->deviceId] -=
+                    memory_record->bytes;
+              }
+              break;
+            }
+            default:
+              LOG_WARNING << "Unrecognized type of memory is allocated, kind "
+                          << memory_record->memoryKind;
+              usage->good_record_ = false;
+              break;
+          }
+        }
+      }
+      break;
+    }
+    case CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION: {
+      CUpti_ActivityExternalCorrelation* corr =
+          (CUpti_ActivityExternalCorrelation*)record;
+      std::lock_guard<std::mutex> lk(mtx_);
```

Comment:
can we add a check for externalcorrelationkind ==CUPTI_EXTERNAL_CORRELATION_KIND_UNKNOWN